### PR TITLE
Pass the Application Note to the BE and save to the DB

### LINKED ISSUE #3020: Status Tracking: Data modeling for status changes
Add a general system event log to store status changes and their related optional emails being sent.

The final sketch for the general system with these two specific first examples is :

```java
public class ApplicationEvents extends BaseModel {
  @ManyToOne private Application application;
  @Constraints.Required private ApplicationEventType eventType;
  // The Account that triggered the event.
  private Account actor;
  // eventType specific details of the event.
  private ApplicationEventDetails details;
  @WhenCreated private Instant createTime;
}

< Index on application id and eventType >

public enum ApplicationEventType {
  STATUS_CHANGE("status_change"),
  NOTE_CHANGE("note_change");
  
  private final String eventType;
}

// A single Application event and its details.
//
// This class is intended to represent the relevant details of an Application
// eventType, as such only one of the Optional detail fields should be present.
// This class will be json serialized into a database column.
public class ApplicationEventDetails {
  // The event affecting the application.
  private final String eventType
  
  Optional<ApplicationStatusChange> applicationStatusChange;
  Optional<ApplicationNoteChange> applicationNoteChange;

  public class ApplicationStatusEvent {
    // The text of the StatusDefinitions.Status applied to the application.
    private String statusText;
    // If the status's email was sent.
    private Boolean emailSent;
  }

  public class ApplicationNoteChange{
    // A note set on the application.
    String note;
  }
}
```

### LINKED ISSUE #3264: Status Tracking: Receive Note Change FE request and save to DB
Send the note from the FE to the BE and save it to the DB.

Sending the request is soon to be done.
The Repository is present, but need the Service and and Controller hookup

### PR DESCRIPTION
**Pull request by shanemc-goog**
### Description

Save the Application Note into the database.

Add an Id and name to the existing UI elements and have the Controller/Service save it as an ApplicationEvent.

## Release notes:

Program Application Notes are now saved into the database. 

### Checklist

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)

Addresses #3020 #3264

### CODE REVIEW

File: server/app/controllers/admin/AdminApplicationController.java [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 import static com.google.common.base.Preconditions.checkNotNull;
 import static views.admin.programs.ProgramApplicationView.NEW_STATUS;
+import static views.admin.programs.ProgramApplicationView.NOTE;
```
Comment (clouser2) [inline server/app/controllers/admin/AdminApplicationController.java:5]: Leaving comment here for general PR description: This doesn't address all of the open TODOs in #3020, so I'd recommend not marking it as fixing the issue. For example, the changes in PR #3306 will be required first.

In that PR, I repointed the "show the current note" TODOs against #3264, so I'd also suggest not marking this as fixing that issue. I couldn't find an issue in the status tracking tracker that was along the lines of "show the current note" that would be appropriate to point it against. I'd honestly just use #3264 for that, since setting an application note and being able to observe the changes seems like the smallest useful unit for a user.

Totally fine as a separate PR if you'd prefer, I'd just suggest not marking this PR as fixing the issues since there's still remaining work.

File: server/app/services/applications/ProgramAdminApplicationService.java [hunk 2]
```diff
@@ -58,4 +64,20 @@ public void setStatus(Application application, StatusEvent newStatus, Account ad
             application, admin, ApplicationEventDetails.Type.STATUS_CHANGE, details);
     eventRepository.insertSync(event);
   }
+
+  /**
+   * Saves a new {@link ApplicationEventDetails.Type.NOTE_CHANGE} event.
+   *
+   * @param admin The Account that instigated the change.
+   */
+  public void setNote(Application application, NoteEvent note, Account admin) {
```
Comment (clouser2) [inline server/app/services/applications/ProgramAdminApplicationService.java:73]: Not blocking: Reading the docstring next to the method name is a bit incongrous (docstring=Saves a new and methodname=setNote). setNote implies there's a single note, while we model it as a new event of a given type. The UI will only display the latest note (at least as I understand). That said, at the service level, it feels like we're tying the method name to that implementation a little too strongly.

File: server/app/services/applications/ProgramAdminApplicationService.java [hunk 3]
```diff
@@ -58,4 +64,20 @@ public void setStatus(Application application, StatusEvent newStatus, Account ad
...
+    ApplicationEventDetails details =
+        ApplicationEventDetails.builder()
+            .setEventType(ApplicationEventDetails.Type.NOTE_CHANGE)
+            .setNoteEvent(note)
+            .build();
+    ApplicationEvent event =
+        new ApplicationEvent(application, admin, ApplicationEventDetails.Type.NOTE_CHANGE, details);
```
Comment (clouser2) [inline server/app/services/applications/ProgramAdminApplicationService.java:80]: Didn't notice originally, but why do we have to specify the ApplicationEventDetails.Type both in the ApplicationEventDetails builder as well as the ApplicationEvent constructor? Could we remove the argument from the constructor and instead infer it from the provided details? That would cut out a class of potential mismatch errors that might come our way and simplify usage.

File: server/app/views/admin/programs/ProgramApplicationView.java [hunk 4]
```diff
@@ -269,17 +270,24 @@ private Modal renderUpdateNoteConfirmationModal(
       long programId, Application application, Http.Request request) {
     ButtonTag triggerButton =
         makeSvgTextButton("Edit note", Icons.EDIT).withClasses(AdminStyles.TERTIARY_BUTTON_STYLES);
+    // Give each form a unique id based on the application.
+    String formId = String.format("note-form-%d", application.id);
```
Comment (clouser2) [inline server/app/views/admin/programs/ProgramApplicationView.java:274]: If we're not selecting based on this ID anywhere, trying to make it human readable doesn't provide a lot of benefit in my opinion. It also makes it potentially more likely to run into conflicts if we were to either 1) copy/paste or 2) Have a similar element with the same name elsewhere in the DOM. Both of these are unlikely, obviously.

That said, if it doesn't need to have human-readable meaning, I'd prefer making that explicit through initializing the form with a random ID.

We could use a random UUID. It'd make sense in my opinion to pull Modal.randomModalId() into a helper function in something in app/views for this purpose.

File: server/test/controllers/admin/AdminApplicationControllerTest.java [hunk 5]
```diff
@@ -363,6 +363,55 @@ public void updateNote_notAdmin() throws Exception {
...
+    // Execute.
+    Result result = controller.updateNote(request, program.id, application.id);
+
+    // Verify.
+    assertThat(result.status()).isEqualTo(BAD_REQUEST);
+    assertThat(contentAsString(result)).contains("A note is not present");
+  }
+
+  @Test
+  public void updateNote_succeeds() throws Exception {
```
Comment (clouser2) [inline server/test/controllers/admin/AdminApplicationControllerTest.java:386]: Mind adding a test with a note of the empty string "" succeeding? I think supporting this is definitely desirable since it's effectively "clearing" any existing notes. I just think it'll be helpful to have a test case that confirms explicit support for that value (and that it's not, for example, treated as a bad request).

**Review by clouser2 — CHANGES_REQUESTED** [review]

File: server/app/controllers/admin/AdminApplicationController.java (hunk 1 above)
Comment (shanemc-goog) [reply server/app/controllers/admin/AdminApplicationController.java:5]: I don't believe having Fixes closes the Issue, It's more as "This addresses issue X".

AFAIK our Fixes section is just self-imposed organization we have created similar to the Release Notes section and the checklist.  It's meaningless aside from humans who read it.

File: server/app/services/applications/ProgramAdminApplicationService.java (hunk 3 above)
Comment (shanemc-goog) [reply server/app/services/applications/ProgramAdminApplicationService.java:80]: They both have different utility, one is at the database level to query on, the other is that I think it's reasonable to want to pass the Details object rather than the entire Event.  I might be wrong there, but felt better to start with than add subsequently.

File: server/app/controllers/admin/AdminApplicationController.java (hunk 1 above)
Comment (clouser2) [reply server/app/controllers/admin/AdminApplicationController.java:5]: I don't think that's quite correct. Empirically, GitHub closes issues prefixed with the "Fixes" keyword, I've made use of this in many PRs. To just reference another issue, not prefixing with "Fixes" does the trick (https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

I've observed this behavior with PRs. See https://github.com/civiform/civiform/pull/3289 for an example where the issue was automatically marked as fixed once the merge had completed and https://github.com/civiform/civiform/pull/3207 for an example where the issue didn't have the "Fixes" prefix and was still referenced (but not closed).

File: server/app/services/applications/ProgramAdminApplicationService.java (hunk 3 above)
Comment (clouser2) [reply server/app/services/applications/ProgramAdminApplicationService.java:80]: Put another way, is there ever a case where we expect the two to mismatch? If not, then I don't see the need for it in the constructor. EBean's population of an ApplicationEvent won't use the constructor, I don't believe. It will instead use the provided getters/setters. Constructor initialization is only for our application code, where I don't think there's a need to specify them separately.

File: server/app/services/applications/ProgramAdminApplicationService.java (hunk 2 above)
Comment (shanemc-goog) [reply server/app/services/applications/ProgramAdminApplicationService.java:73]: At some layer we should be imposing business logic on top of the data and I think that's appropriate in the Service layer as it's the lowest non database oriented layer.

There is logically 1 active status or note, there are also historical ones that the Service may later reveal to the a Controller.

As you've brought up before, we don't have great lines as such, often using the data class directly in controllers, but I don't think that is a good thing to not try to impost the business logic of viewing the database in a Service.

Thoughts?

File: server/app/controllers/admin/AdminApplicationController.java (hunk 1 above)
Comment (shanemc-goog) [reply server/app/controllers/admin/AdminApplicationController.java:5]: Ah confusing.  Seemingly you can't get it right if two parallel PRs address an Issue and collectively fix it.

updated

File: server/app/services/applications/ProgramAdminApplicationService.java (hunk 3 above)
Comment (shanemc-goog) [reply server/app/services/applications/ProgramAdminApplicationService.java]: Ah sorry, I missed the second infer part of your original comment, remove the param in the ctor.

File: server/app/views/admin/programs/ProgramApplicationView.java (hunk 4 above)
Comment (shanemc-goog) [reply server/app/views/admin/programs/ProgramApplicationView.java]: Done. Modal is already here so used it directly.

File: server/test/controllers/admin/AdminApplicationControllerTest.java (hunk 5 above)
Comment (shanemc-goog) [reply server/test/controllers/admin/AdminApplicationControllerTest.java:386]: agreed and done!

**Review by clouser2 — APPROVED** [review]
Looks good!

File: server/app/services/applications/ProgramAdminApplicationService.java (hunk 2 above)
Comment (shanemc-goog) [reply server/app/services/applications/ProgramAdminApplicationService.java:71]: Hmm, that's a good approach, at least we can _try_ to enforce business logic through methods and their docs.

I'd prefer to do that in a separate PR though.
